core/src/com/unciv/logic/map/MapGenerator.kt

```diff
@@ -371,15 +372,66 @@ class MapGenerator(val ruleset: Ruleset) {
 
     // Here, we need each specific resource to be spread over the map - it matters less if specific resources are near each other
     private fun spreadStrategicResources(tileMap: TileMap, distance: Int) {
-        val strategicResources = ruleset.tileResources.values.filter { it.resourceType == ResourceType.Strategic }
-        // passable land tiles (no mountains, no wonders) without resources yet
-        val candidateTiles = tileMap.values.filter { it.resource == null && it.isLand && !it.getLastTerrain().impassable }
-        val totalNumberOfResources = candidateTiles.size * tileMap.mapParameters.resourceRichness
-        val resourcesPerType = (totalNumberOfResources/strategicResources.size).toInt()
+        // Extensive but fast preparations - not many terrains therefore all collections are very small
+        // What, including terrain features, can count as land or water, respectively?
+        // This would make little sense with the vanilla ruleset but with this we allow some pretty free modding
+        val baseLand = ruleset.terrains.values.asSequence().filter { it.type == TerrainType.Land }
+        val landFeatures = ruleset.terrains.values.asSequence().filter { it.type == TerrainType.TerrainFeature && it.occursOn != null && it.occursOn.any { it in baseLand.map { it.name } } }
+        val allLand = (baseLand+landFeatures).map { it.name }.toHashSet()
+        val baseWater = ruleset.terrains.values.asSequence().filter { it.type == TerrainType.Water }
+        val waterFeatures = ruleset.terrains.values.asSequence().filter { it.type == TerrainType.TerrainFeature && it.occursOn != null && it.occursOn.any { it in baseWater.map { it.name } } }
+        val allWater = (baseWater+waterFeatures).map { it.name }.toHashSet()
+
+        // Now, per resource: Can they occur on land and/or water?
+        data class LandOrWater(val land: Boolean = false, val water: Boolean = false)
+        fun getLandOrWater(resource: TileResource): LandOrWater {
+            return LandOrWater(resource.terrainsCanBeFoundOn.any { it in allLand }, resource.terrainsCanBeFoundOn.any { it in allWater} )
+        }
+        val strategicResources =
+                ruleset.tileResources.values
+                .filter { it.resourceType == ResourceType.Strategic }
+        val landOrWaterMap =
+                strategicResources
+                .map { Pair<String,LandOrWater>(it.name, getLandOrWater(it)) }
+                .toMap()
```

Comment:
I'm pretty sure there's a toMap() that accepts a key and value selector?

Reply:
There should be, but me or the IDE can't find it. Plus, in this part overhead is irrelevant - we're building a six-string to twelve-bools map here?

Reply:
`buildMap<String,LandOrWater> { strategicResources.forEach { put (it.name, getLandOrWater(it)) }  }` would need _experimental_ annotation - and isn't much more readable.